src/shared/sol-util.c

```diff
@@ -905,3 +905,79 @@ sol_util_uuid_str_valid(const char *str)
 
     return true;
 }
+
+SOL_API int
+sol_util_escape_quotes(const struct sol_str_slice slice,
+    struct sol_buffer *buf)
+{
+    bool is_escaped = false;
+    size_t i, last_append;
+    int r;
+
+    SOL_NULL_CHECK(buf, -EINVAL);
+
+    sol_buffer_init(buf);
+
+    if (!slice.len)
+        return 0;
+
+    last_append = 0;
+
+    for (i = 0; i < slice.len; i++) {
+        if (!is_escaped && (slice.data[i] == '"' || slice.data[i] == '\'')) {
```

Comment:
this is not right. If we have quotes, then we should adapt the indexes and pointers so if there is no escape, we can fallback to the `if (!last_append)` case you handle post the loop is finished, that is: no copy, just return a slice of the input buffer without the quotes.

Given the usage I'd handle the following input as tests (please add a test case for those):
 - `x` -> `x`, no copy
 - ` x` -> ` x`, no copy (leading space, keep it)
 - `x ` -> `x `, no copy (trailing space, keep it)
 - `"x"` -> `x`, no copy
 - ` "x"` -> `x`, no copy (leading space, ignore it)
 - `"x" ` -> `x`, no copy (trailing space, ignore it)
 - `"x'` -> error
 - `'x"` -> error
 - `"x` -> error
 - `'x` -> error
 - `x"` -> error
 - `x'` -> error
 - `x"y"` -> `xy` (or error, or copy verbatim, your pick, but document it).
 - `x'y'` -> `xy` (or error, or copy verbatim, your pick, but document it).
 - `"x\"y"` -> `x"y`, copy/unescape
 - `"x\'y"` -> `x'y`, copy/unescape

Also add test to check if it was NOT copied, we need to check if we are handling the optimized cases right.

anyway, as you can see we need to change this loop